pkg/core/resources/apis/meshservice/match_benchmark_test.go

```diff
@@ -0,0 +1,88 @@
+package meshservice_test
+
+import (
+	"fmt"
+	"testing"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/meshservice"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/meshservice/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/test/resources/builders"
+)
+
+// To run, remove X from the prefix
+// $ cd /Users/jakub/kong/kuma/pkg/core/resources/apis/meshservice
+// $ go test -bench=BenchmarkMatchDataplanesWithMeshServices -run=^# -count 10 | tee bench.txt
```

Comment:
have you checked with `-benchmem`? having maps with index probably consumes more memory

Reply:
it's... strange?

```
❯❯❯ go test -bench=BenchmarkMatchDataplanesWithMeshServices -benchmem -run=^# -count 10
goos: darwin
goarch: arm64
pkg: github.com/kumahq/kuma/pkg/core/resources/apis/meshservice
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02280 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02266 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02296 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02321 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02315 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02241 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02327 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02239 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02266 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10         	1000000000	         0.02324 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5791806292 ns/op	 1121608 B/op	   16090 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5666001125 ns/op	 1120456 B/op	   16086 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5618513709 ns/op	 1121896 B/op	   16091 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5781507583 ns/op	 1120744 B/op	   16087 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5621502292 ns/op	 1129672 B/op	   16118 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5686973541 ns/op	 1123336 B/op	   16096 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5633431500 ns/op	 1123336 B/op	   16096 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5632041709 ns/op	 1120168 B/op	   16085 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5753600167 ns/op	 1120744 B/op	   16087 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10          	       1	5664713792 ns/op	 1118728 B/op	   16080 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001502 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001499 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001478 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001551 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001476 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001514 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001439 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001474 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001425 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10           	1000000000	         0.001485 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.01988 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02029 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02034 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02035 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02048 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02073 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02023 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02028 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02041 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10            	1000000000	         0.02128 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001132 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001112 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001227 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001148 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001202 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001103 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0002870 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001277 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001054 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10           	1000000000	         0.0001173 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001843 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001810 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001695 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001712 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001816 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001729 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001656 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001684 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001814 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10            	1000000000	         0.0001744 ns/op	       0 B/op	       0 allocs/op
```

not sure why it's zero. I mean, for sure memory usage is higher, but I think it's worth it. The previous algo just does not scale.

Reply:
ah, I think you're missing loops from 0 to `b.N`:
```go
for i := 0; i < b.N; i++ {
    _ = meshservice.MatchDataplanesWithMeshServices(dpps, meshServices, false)
}
```

When I added this I can see allocs and mem usage:
```
➤ benchstat bench.txt
goos: darwin
goarch: arm64
pkg: github.com/kumahq/kuma/pkg/core/resources/apis/meshservice
                                                                       │  bench.txt  │
                                                                       │   sec/op    │
MatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10   23.96m ± 4%
MatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10     5.243 ± 4%
MatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10     1.447m ± 4%
MatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10      18.51m ± 0%
MatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10     89.15µ ± 1%
MatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10      149.5µ ± 2%
geomean                                                                  5.961m

                                                                       │  bench.txt   │
                                                                       │     B/op     │
MatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10   15.61Mi ± 0%
MatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10    1.055Mi ± 0%
MatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10     971.0Ki ± 0%
MatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10      67.88Ki ± 0%
MatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10     84.60Ki ± 0%
MatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10      9.911Ki ± 0%
geomean                                                                  313.8Ki

                                                                       │  bench.txt  │
                                                                       │  allocs/op  │
MatchDataplanesWithMeshServices/faster-matching-ns_200-svc_20-dpp_5-10   370.5k ± 0%
MatchDataplanesWithMeshServices/naive-matching-ns_200-svc_20-dpp_5-10    16.04k ± 0%
MatchDataplanesWithMeshServices/faster-matching-ns_50-svc_5-dpp_5-10     23.34k ± 0%
MatchDataplanesWithMeshServices/naive-matching-ns_50-svc_5-dpp_5-10      1.014k ± 0%
MatchDataplanesWithMeshServices/faster-matching-ns_10-svc_5-dpp_1-10     1.497k ± 0%
MatchDataplanesWithMeshServices/naive-matching-ns_10-svc_5-dpp_1-10       58.00 ± 0%
geomean                                                                  4.799k
```

Reply:
thanks, it was the missing loop indeed. So yes, the memory usage is higher, but still 23ms vs 5.2s on 4k services and 20k DPPs